text/0249-v2-experiments.md

```diff
@@ -0,0 +1,137 @@
+  implemented in a separate package. This might lead to awkward choices, like making functions public that otherwise
+  would have been private.
+
+# Alternatives
+
+Some alternatives to this strategy are:
+
+- Make no changes and continue to experiment in the main release.
+  - Improve documentation so that it is more obvious to users that they are importing experimental code.
+  - Make it a policy to actually include the name 'experimental' in all experimental APIs.
+- Use clever build tricks to hide experimental modules. This might be possible in Typescript, where the type files are
+  separate from implementation, but might prove difficult or impossible in other languages such as Python.
+- Call the experimental package @aws-cdk-experiments instead of @aws-cdk-previews.
+
+# Adoption Strategy
+
+The hope is that for the majority of users, upgrading from 1.x to 2.x will be as simple as changing import statements at
+the top of their source files, rebuilding, and redeploying. This upgrade should not result in any resources being
+destroyed and recreated, except in very specific cases where we have made it clear to users, such as with the EKS L2
+construct library. The same should be true for users upgrading from the preview package to the stable package as modules
+are graduated from Developer Preview to GA.
+
+# Implementation Plan
+
+This section contains examples covering common scenarios:
+
+## 1. A brand new L2, where there was only an L1 before.
+
+- Create a new package in packages/@aws-cdk-previews exactly the same we we would have for v1
+- Follow the v1 process for module lifecycle until it's ready to GA
+- [How does documentation work for the preview package?]
+- To graduate it, move the package to packages/@aws-cdk, set the module stability and maturity to stable, and remove any
+  @experimental tags in the doc comments.
+
+## 2. Adding new experimental classes to an existing L2.
+
+An example would be Cognito User Pools, when we want to add additional Identity Pool Providers.
+
+- Create the implementation file and export it.
+  - packages/@aws-cdk-previews/cognito/lib/user-pool-idps/OpenIdConnect.ts
+- A user that wants to try it would import both the stable and experimental package
+
+```typescript
+import * as cognito from '@aws-cdk-lib/aws_cognito';
+import * as cognitoPreview from '@aws-cdk-previews/aws_cognito';
+
+const idp = new cognitoPreview.UserPoolIdentityProviderOidc(this, 'OIDC', {...});
+const supported = [cognito.UserPoolClientIdentityProvider.custom("MyProviderName")];
+const userPoolClient = new cognito.UserPoolClient(...);
+
+```
+
+## 3. Adding a new experimental function to an existing stable class.
+
+TODO
+
+## 4. How to work around stable modules that want to refer to experiments
+
+TODO - stepfunctions-tasks?
+
+## 5. Changing the entire implementation for a core module.
+
+TODO - asset bundling?
```

Comment:
?

Reply:
I need someone to help me with these sections, since I don't have enough experience with implementation.

Reply:
Okay we need to dive deeper into these sections to see the full picture